### ISSUE TITLE
How to whitelist (permit user) to a single directory?

### ISSUE THREAD

**Issue by danielkrajnik** [netblue30/firejail#5879]
Whitelisting a single directory causes other directories to be blacklisted. Is there a method that simply whitelists a directory/file without any side effects?

Sorry if this has been asked before, but I couldn't find any answer to that (not sure how would you find it). I realize that there is a template for this type of issues, but it's a really simple question and I would be really grateful for your help.

**Comment by glitsj16** [comment]
> Whitelisting a single directory causes other directories to be blacklisted. Is there a method that simply whitelists a directory/file without any side effects?

If you're asking if there is a kind of whitelisting without blacklisting other directories the answer is no. That's what whitelisting is all about. Perhaps we can offer more help if you provide an example of what you're trying to achieve exactly.

**Comment by danielkrajnik** [comment]
thunderbird requires access to a directory that stores an .ics file (a calendar). This directory is a symlink in home (`notes`) to a folder in /mnt (`/mnt/notes`), which has user permissions. However, when default firejail thunderbird.profile is applied it doesn't see this directory.

I've added thunderbird.local to .config/firejail:
````
noblacklist ~/notes
whitelist ~/notes

noblacklist /mnt/notes
# whitelist /mnt/notes
```
however this doesn't work - you can see notes folder from thunderbird, but access is denied.

If you uncomment the ``whitelist /mnt`/notes` thunderbird doesn't start at all - I assume that's because whiteslist implicitly blacklists everything else?

Is there a simple solution to what I thought is a very common use case - whitelisting a folder that user has normally access to. 

Admittedly it's in `/mnt`  location, but permissions are the same.

**Comment by glitsj16** [comment]
> Thank you for not closing this issuing straight away by the way, I imagine that this same question must be asked all the time and it's probably answered somewhere in the documentation already.

No problem. This isn't all that straightforward if one isn't yet very familiar with the Firejail profiles and the rather large set of options. The reason why your whitelisting isn't working can be found if you follow the include logic. In this case our thunderbird.profile includes firefox-common.profile, which has the option `disable-mnt`:

https://github.com/netblue30/firejail/blob/33c75b89328df03ef3245c7ec6f30759f9619223/etc/profile-a-l/firefox-common.profile#L58

So you need to `ignore disable-mnt` in your thunderbird.local and then whitelisting ${HOME}/notes and /mnt/notes should work. No need for any noblacklist. That option only makes sense to override a previously blacklisted path, which isn't done anywhere for ~/notes in any of our profiles.

**Comment by danielkrajnik** [comment]
Sorry, but this doesn't seem to work:

thunderbird.local:
````
ignore disable-mnt
whitelist ${HOME}/notes
whitelist /mnt/notes
```
$ thunderbird
````
Reading profile /etc/firejail/thunderbird.profile
Reading profile /home/user1/.config/firejail/thunderbird.local
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 1069955, child pid 1069974
Warning: NVIDIA card detected, nogroups command ignored
Warning: cleaning all supplementary groups
Warning: NVIDIA card detected, nogroups command ignored
Warning: cleaning all supplementary groups
Warning: NVIDIA card detected, nogroups command ignored
Warning: cleaning all supplementary groups
Error fstat: fs_whitelist.c:156 whitelist_file: Permission denied
Error: proc 1069955 cannot sync with peer: unexpected EOF
Peer 1069974 unexpectedly exited with status 1
````

**Comment by danielkrajnik** [comment]
> This isn't all that straightforward if one isn't yet very familiar with the Firejail profiles

Besides this particular issue, I'm wondering if we would benefit from "global overrides" instead of trying to address specific profiles that . These problems often come up in the middle of your day when you're focused on something else than solving firejail sandbox, which forces you to remove firejail altogether and deal with it later, by the time it may be too late and your previous data is gone. 

Wouldn't it make sense to provide simpler overrides instead?

**Comment by glitsj16** [comment]
This works for me with only whitelisting ~/notes. Who owns your ~/notes and what are the file permissions on /mnt/notes/foo.isc?

```shell
$ ls -l /mnt/notes
total 4
-rw-r--r-- 1 root root 4 Jul  4 01:07 test.txt
```

```shell
$ cat /mnt/notes/test.txt
zzz
```

```shell
$ firejail --quiet --ignore=disable-mnt --whitelist=~/notes --profile=thunderbird cat ~/notes/test.txt
zzz
```

**Comment by danielkrajnik** [comment]
Hmm... that's strange.

````
$ /bin/ls -l /mnt/notes/org-agenda.ics

-rw------- 1 user1 user1 4299 Jul  3 23:16 /mnt/notes/org-agenda.ics
````
```
$ firejail --quiet --ignore=disable-mnt --whitelist=~/notes --profile=thunderbird cat ~/notes/org-agenda.ics

Error fstat: fs_whitelist.c:156 whitelist_file: Permission denied
Error: proc 1316532 cannot sync with peer: unexpected EOF
Peer 1316549 unexpectedly exited with status 1
````

**Comment by rusty-snake** [comment]
Who owns the notes directory.

Does it work if you comment the whitelist if the symlink.

**Comment by glitsj16** [comment]
There's nothing in Firejail that dislikes Mozilla software AFAICT. Honestly, I think there's something off with your permissions that's causing this (and the referenced librewolf) issue. If you cannot fix this by using default dir/file permissions, you might want to write a small shell script that temporarily moves this `org-agenda.ics` out of /mnt/notes BEFORE starting thunderbird and copies it back afterwards. Placing it in either ${HOME}/.cache/thunderbird, ${HOME}/.thunderbird or ${DOWNLOADS} shouldn't need any additional options in a thunderbird.local as these paths are whitelisted in thunderbird.profile by default.

**Comment by danielkrajnik** [comment]
@rusty-snake 

> Who owns the notes directory.

both directory and symlink are owned by the user

```
$ /bin/ls -ld /mnt/notes
drwx------ 1 user1 user1 0 Jul  3 17:35 /mnt/notes 

$ /bin/ls -ld ~/notes
lrwxrwxrwx 1 user1 user1 10 Jun 23 00:12 /home/user1/notes -> /mnt/notes
```

I've tried  commenting out the symlink, but then thunderbird won't start. If you comment out both whitelists it starts but doesn't see that folder.

**Comment by danielkrajnik** [comment]
> There's nothing in Firejail that dislikes Mozilla software AFAICT

Sorry, that was just my poor attempt at humour (it just happened that both mozilla products failed around same time, but these are complex programs after all).

**Comment by pedrib** [comment]
Sorry guys don't want to hijack this thread, but I am running some experiments which are very similar to what @danielkrajnik is attempting to do.

1. I have `disable-mnt` in my globals
2. I want a jailed program to be able to access `/mnt`

Using `--ignore=disable-mnt` on the command line works... but adding `ignore disable-mnt` to the profile doesn't. Is this expected behaviour?

**Comment by glitsj16** [comment]
@pedrib

> Using --ignore=disable-mnt on the command line works... but adding ignore disable-mnt to the profile doesn't. Is this expected behaviour?

No, expected behaviour is that an option works both from command line and from inside a profile. If you have disable-mnt in `globals.local` you need to place an override for it `BEFORE` that file is included. Placing it in a `foo.local` override file should automatically take care of this as those get included first, before globals.local. I've been using such a setup for quite a long time myself and it always worked/never broke. Please open a separate issue with code examples of what is and isn't working as expected.

**Comment by danielkrajnik** [comment]
This is really frustrating, noblacklist option does absolutely nothing. Programs randomly complain about permissions and there is nothing you can do to configure firejail.

**Comment by rusty-snake** [comment]
Simply start with the way it is not accessible. EACCES directs to blacklist options, while ENOENT directs to whitelist options.

**Comment by danielkrajnik** [comment]
Thanks, that's really helpful. Yes, the error was ENOENT - but if you white list that path it will break everything else in that directory. Is there by any chance any way to whitelist only that single directory?

**Comment by danielkrajnik** [comment]
Solved:
````
ignore disable-mnt
whitelist ${HOME}/notes
````

This is exactly what [glitsj16](https://github.com/glitsj16) mentioned in his second comment. I had extra commands there that cause thunderbird to not start at all.

However solving it required familiarizing yourself for a day with firejail manual. The most important moment for me (besides find out the `ignore disable-mnt` option) was finding out that you can inspect your sandbox VERY QUICKLY by going into shell with its current profile (in this instance it's` firejail --profile=thunderbird`). This made a huge difference in iterating and finding out which options are working and which don't.

One thing, which would have made this process a lot easier I think for anyone without having to put so much effort in learning firejail is showing clearly which **profile** causes any certain directory to be blacklisted/read-only/temporary (for example in the output of `firejail --debug` command or in firetools).
